modules/ingester/instance_search.go

```diff
@@ -32,11 +34,47 @@ func (i *instance) Search(ctx context.Context, req *tempopb.SearchRequest) (*tem
 		maxResults = 20
 	}
 
+	span.LogFields(ot_log.String("SearchRequest", req.String()))
 	if api.IsTraceQLQuery(req) {
-		// TODO actually implement recent traceQL search
+		// A trace can be live, in headBlock, completing blocks and complete blocks.
+		// search in this order: HeadBlock, CompletingBlocks, and then finally in completeBlocks.
+		// Iterate over blocks, execute TraceQL, collect results, dedupe, sort and return traces.
+		// return early if we hit maxResults during search.
+		traces, metrics, blockErrs := i.searchWALWithTraceQL(ctx, req, maxResults)
+		err := multierr.Combine(blockErrs...)
+		if err != nil {
+			span.LogFields(ot_log.Error(fmt.Errorf("blockErrors: %w", err)))
+		}
```

Comment:
Not sure about the error handling in search in general. We always return successful, even if we got no results and everything failed. I think we should bubble it up more than a log in a span, that will stay hidden 99% of the time.